app/Http/Controllers/Device/Tabs/NeighboursController.php

```diff
@@ -52,6 +55,86 @@ public function name(): string
 
     public function data(Device $device): array
     {
-        return [];
+        $selection = Url::parseOptions('selection', 'list');
+
+        $links = [];
+        $devices = [];
+
+        if ($selection == 'list') {
+            $linkQuery = Port::select(
+                \DB::raw('ports.*'),
+                \DB::raw('l.remote_port_id'),
+                \DB::raw('l.remote_hostname'),
+                \DB::raw('l.remote_port'),
+                \DB::raw('l.remote_platform'),
+                \DB::raw('l.protocol'),
+            )
+                ->with('device')
+                ->join('links as l', 'l.local_port_id', '=', 'ports.port_id')
+                ->where('l.local_device_id', '=', $device->device_id)
+                ->orderBy('ports.ifName');
+
+            // Only show where user has access to both devices
+            if (! \Auth::user()->hasGlobalRead()) {
+                $linkQuery->whereIntegerInRaw('l.local_device_id', \Permissions::devicesForUser())
+                    ->whereIntegerInRaw('l.remote_device_id', \Permissions::devicesForUser());
+            }
+
+            foreach ($linkQuery->get() as $port) {
+                $row = json_decode(json_encode($port));
+
+                if (! in_array($port->device->device_id, $devices)) {
+                    $devices[$port->device->device_id] = [
+                        'url'  => Url::deviceLink($port->device, null, [], 0, 0, 0, 1),
+                        'hw'   => $port->device->hardware,
+                        'name' => $port->device->shortDisplayName(),
+                    ];
+                }
+
+                $rport = null;
+                if ($row->remote_port_id) {
+                    $rport = Port::where('port_id', '=', $row->remote_port_id)->with('device')->first();
```

Comment:
This create n+1 queries. You should include it in big query; https://laravel.com/docs/10.x/eloquent-relationships#nested-eager-loading

Reply:
Please note that the existing page that this replaces (includes/html/pages/device/ports/neighbours.inc.php) calls device_by_id_cache() once for each neighbour, so this new page will have the same number of SQL queries as the original page.  The ideal solution would be to add links to the DB model, but I can't see why replacing a page with another that has the same number of queries would be an issue if the better solution is too tricky.

Reply:
device_by_id_cache() calls DeviceCache::get() in the background, which will only fetch each device once.
The best way would be to add a relationship to the Ports model to represent neighbor ports.